File: src/com/fictional_particle_sim/geometrics/Point.java
package com.fictional_particle_sim.geometrics;

public class Point {
    public double x, y;
    public Point() {
        x = 0;
        y = 0;
    }
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public Point add(Point b) {

        return new Point(this.x + b.x, this.y + b.y);
    }

    public Point sub(Point b) {

        return new Point(this.x - b.x, this.y - b.y);
    }

    public Point mult(double b) {

        return new Point(this.x * b, this.y * b);
    }

    public double dist(Point b) {

        return Math.hypot(b.x-this.x, b.y-this.y);
    }
    public Vector disp(Vector b) {
        Vector temp = new Vector(b.start, this);
        return temp.perp(b);
    }

    public boolean between(Vector a, Vector b) {
        Vector[] dispA = disp(a).comp();
        Vector[] dispB = disp(b).comp();

        return dispA[0].opposite(dispB[0]) && dispA[1].opposite(dispB[1]);
    }

    public boolean equals(Point b) {

        return this.x == b.x && this.y == b.y;
    }

    @Override
    public String toString() {
        return "( " + x + " , " + y + " )";
    }
}

File: src/com/fictional_particle_sim/geometrics/Point.kt
package com.fictional_particle_sim.geometrics

import kotlin.math.round

fun Point(x: Int, y: Int): Point = Point(x.toDouble(), y.toDouble())
class Point(var x: Double = 0.0, var y: Double = 0.0) {

    operator fun plus(that: Point) = Point(this.x + that.x, this.y + that.y)

    operator fun minus(that: Point) = Point(this.x - that.x, this.y - that.y)

    operator fun times(that: Double) = Point(this.x * that, this.y * that)
    operator fun Double.times(that: Point) = that * this

    operator fun div(that: Double) = this * (1.0/that)

    fun dist(that: Point = Point()) = Vector(this, that).magnitude()
    fun disp(that: Vector) = Vector(that.start, this).perp(that)

    fun inside(k: BoundingBox): Boolean = x <= k.a.x.coerceAtLeast(k.b.x) && x >= k.a.x.coerceAtMost(k.b.x) && y <= k.a.y.coerceAtLeast(k.b.y) && y >= k.a.y.coerceAtMost(k.b.y)

    override fun toString(): String = "( ${round(x * 10000) / 10000} , ${round(x * 10000) / 10000} )"

}